### ISSUE TITLE
CSC crashes to missing System.Memory

### ISSUE THREAD

**Issue by borgdylan** [mono/mono#13735]
## Steps to Reproduce

1. Update to the latest mono from master.
2. Try to compile something using MSBuild that would invoke csc.exe
3. See that System.Memory is needed but unavailable.

### Current Behavior

FIleNotFoundException saying System.Memory 4.0.1.0 could not be found.

### Expected Behavior

A successful compilation.

## On which platforms did you notice this

[ ] macOS
[x ] Linux
[ ] Windows

**Version Used**:

6.1 from master

## Stacktrace (truncated due to length limits)

```
Please paste the stack trace here if available.
```
System.AggregateException: One or more errors occurred. (System.IO.FileNotFoundException: Could not load file or assembly 'System.Memory, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies.
  File name: 'System.Memory, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SlidingTextWindow.Intern (System.Char[] array, System.Int32 start, System.Int32 length) [0x00000] in <d6728d536b884cd2af1698e1ee49045c>:0 
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.ScanIdentifier_FastPath (Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer+TokenInfo& info) [0x00263] in <d6728d536b884cd2af1698e1ee49045c>:0 
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.ScanIdentifier (Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer+TokenInfo& info) [0x00000] in <d6728d536b884cd2af1698e1ee49045c>:0 
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.ScanIdentifierOrKeyword (Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer+TokenInfo& info) [0x00007] in <d6728d536b884cd2af1698e1ee49045c>:0 
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.ScanSyntaxToken (Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer+TokenInfo& info) [0x00870] in <d6728d536b884cd2af1698e1ee49045c>:0 
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.LexSyntaxToken () [0x0003b] in <d6728d536b884cd2af1698e1ee49045c>:0 
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.Lex (Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LexerMode mode) [0x00024] in <d6728d536b884cd2af1698e1ee49045c>:0 
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxParser.PreLex () [0x00047] in <d6728d536b884cd2af1698e1ee49045c>:0 
    at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxParser..ctor (Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer lexer, Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LexerMode mode, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode oldTree, System.Collections.Generic.IEnumerable`1[T] changes, System.Boolean allowModeReset, System.Boolean preLexIfNotIncremental, System.Threading.CancellationToken cancellationToken) [0x00094] in <d6728d536b884cd2af1698e1ee49045c>:0 
  
```
<!--
You can join us at https://gitter.im/mono/mono to discuss your reported issue
-->

**Comment by kg** [comment]
Is it possible for you to invoke mono with --debug so we get filenames and line numbers?

**Comment by kg** [comment]
Also, is it possible to get any information on the test .cs file(s) you were compiling? I cannot reproduce this yet but I am sure I missed an important step. It would be good to know what command line you used to invoke msbuild as well.

**Comment by borgdylan** [comment]
Msbuild was invoked without any parameters. The code is covered by NDA, so I cannot disclose it. I am using msbuild compiled from source and mono compiled from source on x64 Linux. When I looked in the installation folders I could not find System.Memory or related facades under /usr/local/lib/mono/4.5. Thats the directory csc.exe is installed in. The installation was done using the standard ```sudo make install```

**Comment by borgdylan** [comment]
To summarise, my mono compilation process is:
```
git pull origin
./autogen.sh
make clean
make
sudo make install
```

**Comment by kg** [comment]
This info helps very much. How did you compile msbuild from source? A simple explanation is enough. (There are multiple ways to do it)

**Comment by borgdylan** [comment]
I got code from: https://github.com/mono/msbuild/commits/xplat-master
```
make all-mono
sudo install-mono-prefix.sh /usr/local
```
```
dylan@zenbook-pro:~$ msbuild /version
Microsoft (R) Build Engine version 16.0.42-preview+g804bde742b for Mono
Copyright (C) Microsoft Corporation. All rights reserved.
```

**Comment by borgdylan** [comment]
I was planning to update my msbuild as well but holding off in case something goes awry and breaks my work machine like what happened when I updated mono itself.

**Comment by kg** [comment]
We have recently updated roslyn in mono/mono master which is why you are having this problem, and msbuild has a dependency on roslyn so I agree that you should not update your msbuild yet.

Thanks for your patience.

**Comment by borgdylan** [comment]
I have since reverted the mono upgrade, by checking out the previosly known commit and installing that.

**Comment by borgdylan** [comment]
@kg I guess this: https://github.com/mono/mono/commit/0de947787c5bbbac667a1a5ad1cc81c82e352cae is related towards a fix. I noticed that System.Buffers.dll (a recursive dependency from System.Memory.dll) mis not included. Was this added as a facade or otherwise in a recent commit?

**Comment by kg** [comment]
@borgdylan yeah, some infrastructure issues prevented us from identifying the full dependency chain. We have facades for some of these libraries but not all, which makes it tricky. We have reverted the roslyn bump so latest master should work for you.

**Comment by borgdylan** [comment]
Thanks a million!
